Dashboard chart widget empty on first mount (pie & bar): data present, repaints on reload — first screen of every AI-built app

## Symptom (P2, recurring): dashboard chart widget renders EMPTY on first mount — data is fine, a refresh paints it

2026-08-21 live repro on the cloud local rig (objectui `fd227eae1`, console built from workspace): open an AI-built dashboard (`客户管理首页仪表盘`, one pie widget bound to a dataset with 3 categories / 12 rows). First navigation into the dashboard renders the widget card with its title and an EMPTY chart area (no slices, no axes, no empty-state message). Press F5 → the pie renders correctly (3 slices).

The same behavior was observed on 2026-08-13 with a bar widget (no bars on mount, data confirmed present via the data API, recovers on re-render) — so it is not chart-type-specific and has survived at least a week of main.

## Why it matters

This is the FIRST screen an AI-build user sees after publishing ("首页仪表盘按状态统计客户数量" is the standard magic-flow prompt). An empty chart on first open reads as "the AI built a broken app" — it undoes the entire publish pipeline's credibility for the cost of one missed re-render.

## Repro

1. cloud rig `run-stack.sh --with-ui`, build an app via the build agent with a dashboard (pie or bar) bound to a real dataset.
2. Publish; from the launcher open the app → dashboard nav entry.
3. First render: empty chart area. Reload: chart paints.

Suspected class: the chart mounts before its dataset query resolves (or before the container has measured dimensions) and never re-draws when data arrives — race, not data. Not bisected; the fix owner should instrument mount/data-arrival ordering in the dashboard widget renderer.

<!-- os-dev-report -->

```json
{
  "issue": 5451,
  "status": "rework",
  "branch": "claude/issue-5451-dashboard-chart-first-mount",
  "pr": "https://github.com/objectstack-ai/objectui/pull/5503",
  "premise_still_valid": false,
  "summary": "premise_still_valid:false refers to the card's MECHANISM, not its symptom — the symptom class is real and this PR fixes one cause of it; the PR is the 'surviving part' the contract asks for, not a full close, hence status:rework and 'Part of #5451' rather than a closing keyword. Instrumented in a real browser (Vite over the console's workspace src/ aliases + Chromium 141 via Playwright) driving DashboardRenderer -> SchemaRenderer -> registry 'chart' -> ChartRenderer -> React.lazy(AdvancedChartImpl) inside a byte-for-byte copy of DashboardView's wrapper markup, plugins registered lazily as apps/console/src/register-plugins.ts does. REFUTED: (a) the mount/data-arrival race — dataset-bound, provider:'object' and dashboard-grid paths all paint on first mount at 0/120/800ms data delay; (b) 'mounts unmeasured and never redraws' — a chart mounted inside display:none shows the empty state and then paints every mark the moment the subtree is revealed, because recharts 3.10's own ResizeObserver re-measures. So a LATE measurement heals itself. What does not heal is a PERMANENT zero, and I found one: ChartContainer's min-size fallback is applied to the wrapper div only, while recharts measures its own width:100%/height:100% div, and a percentage height never resolves against an ancestor's min-height (CSS 2.1 s10.5). Measured: wrapper 638x280 (floor applied, looks healthy) while the measured div is 638x0 -> 0 surfaces, 0 marks, no refusal, no empty state, no loading state. Chart-type-independent, matching the card. Fix forwards the same floor to ResponsiveContainer under the same hasDeclaredHeight gate, so an author's explicit height still wins. No setTimeout, no extra observer, no re-mount. NOT ESTABLISHED: the card's first-mount-vs-reload asymmetry — I could not produce it, so I cannot claim this is the maintainer's exact repro; #5451 should stay open until re-checked on the live rig with this fix. PM assumption check: the locus is packages/plugin-charts, not plugin-dashboard; nothing in plugin-dashboard was touched and #5445's dimension-member label pins were not disturbed. Card body read in full via the API and complete — no sanitizer truncation (the entity escaping in it is API encoding, not damage).",
  "tests": "All at HEAD 3323b4b5f (re-run AFTER the final commit); heavy steps via scripts/pm/os-verify-lock.sh, verdicts quoted from the lock's own VERDICT line. (1) 'pnpm --filter @object-ui/plugin-charts type-check' + 'pnpm exec vitest run packages/plugin-charts/' -> script name echoed (not a zero-match run), 'Test Files 32 passed (32)' / 'Tests 235 passed (235)', 'os-verify-lock: VERDICT command-exit 0'. (2) the five plugin-dashboard tests that import @object-ui/plugin-charts and render the real chart chain (DatasetWidget.chartConfig.dom, DatasetWidget.comboPresentation, DashboardGridLayout.sduiDatasetPath, DashboardGridLayout.datasetPath, StaticTableWidget) -> 'Test Files 5 passed (5)' / 'Tests 46 passed (46)', 'VERDICT command-exit 0'. (3) node scripts/check-changeset-presence.mjs -> '2 source file(s) of 1 released package(s) changed, and this change declares 1 changeset(s)'. (4) node scripts/check-changeset-no-major.mjs -> 'No changeset declares a major bump'. (5) npm run check:control-bytes -> 'check-control-bytes: OK (scanned 4553 tracked text file(s); skipped 85 binary)'. (6) eslint over the whole affected package: 45 files counted from --format json, 0 errors; my two files 0 errors / 0 new warnings (the 6 on ChartContainerImpl.tsx are pre-existing no-explicit-any). Narrowing declared and evidenced: population read from eslint's own config resolution rather than a guess, count read from --format json, and no type-aware linting is configured (no projectService / parserOptions.project anywhere in the eslint configs), so this diff cannot move the verdict on any untouched file. REVERSE VERIFICATION, from a committed state: 'git checkout origin/main -- packages/plugin-charts/src/ChartContainerImpl.tsx', mutation confirmed on disk by grep counts (CHART_MIN_HEIGHT 0 occurrences, original 'minHeight: 280' text present) -> 'Tests 2 failed | 2 passed (4)', 'AssertionError: expected \"\" to be \"280px\"', 'VERDICT command-exit 1'. Predicted direction and observed direction agree: the 2 floor assertions go red while the 2 'author's explicit height wins' assertions stay green, which is required — the fix must not change that precedence. Restored via 'git checkout <branch> -- <path>', restoration confirmed on disk (5 occurrences). No ablation of built dist was involved: the subject is source-level and the harness renders src directly. BROWSER before/after on the same rig, wrapper h-[350px] neutralised so the fallback is the only thing holding the box up — before: measured div 638x0, 0 surfaces, 0 marks; after: measured div 638x280, 2 surfaces, 3 sectors + 3 bars; untouched control unchanged at 638x350 with the same marks both times. DEVIATION (declared): a combined plugin-charts + plugin-dashboard vitest run acquired the lock after 8m48s and was killed by the container's 10-minute foreground cap, so I narrowed plugin-dashboard to the five files that actually import the chart package. The rest of plugin-dashboard is CI's run.",
  "open_questions": [
    {
      "question": "Is the maintainer's live-rig blank chart this permanent-zero-height mode, or a second, still-unfound cause? The measured symptom shape matches exactly (title renders, chart area blank, no message, pie and bar alike, no refusal), but the reported 'F5 paints it' asymmetry is not explained by it — a box that never changes is not fixed by a reload either.",
      "options": [
        "A: re-run the 2026-08-21 cloud-rig repro with PR #5503 applied. If the chart paints, this was the cause and the F5 note was a one-off observation; close #5451. Costs one rig run.",
        "B: keep #5451 open and dispatch a follow-up that captures the failing DOM from the real rig (the measured div's box and the computed height chain on [data-slot=chart]), which is the one datum no synthetic harness can supply. Costs a round but is decisive either way.",
        "C: treat the card as fixed on the strength of the symptom match alone. Cheapest, and wrong if a second cause exists — the symptom would come straight back."
      ],
      "recommendation": "A, then B only if it still reproduces. A is one run and answers the question directly; the fix is already independently justified by the measured before/after, so A risks nothing. C is not acceptable: I refuted the card's stated mechanism but did not reproduce its stated asymmetry, and closing on a symptom match would bury exactly that gap."
    }
  ],
  "out_of_scope_findings": [
    "filed as #5496 (sub-issue of #4695, label 'finding', unassigned, not queued): a chart whose declared series dataKey exists in NO row draws axes around zero marks and says nothing — the third silent-empty-frame arm, covered by neither #4683's 'series: []' guard nor #4695's 'series === undefined' arm. Measured, not inferred: rows carrying {status, count} against series [{dataKey:'id'}] produced 2 surfaces / 0 marks / 0 refusals; correcting the column made both charts paint."
  ]
}
```

---
_Generated by [Claude Code](https://claude.ai/code)_

os-dev-report

(Supersedes the previous comment — its HTML-comment marker was eaten by GitHub's body sanitizer on write, so it is invisible to a marker scan. This one carries the marker as literal text. Same report; two angle-bracket placeholders inside `tests` are also respelled, having been eaten the same way.)

```json
{
  "issue": 5451,
  "status": "rework",
  "branch": "claude/issue-5451-dashboard-chart-first-mount",
  "pr": "https://github.com/objectstack-ai/objectui/pull/5503",
  "premise_still_valid": false,
  "summary": "premise_still_valid:false refers to the card's MECHANISM, not its symptom — the symptom class is real and this PR fixes one cause of it; the PR is the 'surviving part' the contract asks for, not a full close, hence status:rework and 'Part of #5451' rather than a closing keyword. Instrumented in a real browser (Vite over the console's workspace src/ aliases + Chromium 141 via Playwright) driving DashboardRenderer to SchemaRenderer to registry 'chart' to ChartRenderer to React.lazy(AdvancedChartImpl) inside a byte-for-byte copy of DashboardView's wrapper markup, plugins registered lazily as apps/console/src/register-plugins.ts does. REFUTED: (a) the mount/data-arrival race — dataset-bound, provider:'object' and dashboard-grid paths all paint on first mount at 0/120/800ms data delay; (b) 'mounts unmeasured and never redraws' — a chart mounted inside display:none shows the empty state and then paints every mark the moment the subtree is revealed, because recharts 3.10's own ResizeObserver re-measures. So a LATE measurement heals itself. What does not heal is a PERMANENT zero, and I found one: ChartContainer's min-size fallback is applied to the wrapper div only, while recharts measures its own width:100%/height:100% div, and a percentage height never resolves against an ancestor's min-height (CSS 2.1 s10.5). Measured: wrapper 638x280 (floor applied, looks healthy) while the measured div is 638x0, giving 0 surfaces, 0 marks, no refusal, no empty state, no loading state. Chart-type-independent, matching the card. Fix forwards the same floor to ResponsiveContainer under the same hasDeclaredHeight gate, so an author's explicit height still wins. No setTimeout, no extra observer, no re-mount. NOT ESTABLISHED: the card's first-mount-vs-reload asymmetry — I could not produce it, so I cannot claim this is the maintainer's exact repro; #5451 should stay open until re-checked on the live rig with this fix. PM assumption check: the locus is packages/plugin-charts, not plugin-dashboard; nothing in plugin-dashboard was touched and #5445's dimension-member label pins were not disturbed. Card body read in full via the API and complete — no sanitizer truncation (the entity escaping in it is API encoding, not damage).",
  "tests": "All at HEAD 3323b4b5f (re-run AFTER the final commit); heavy steps via scripts/pm/os-verify-lock.sh, verdicts quoted from the lock's own VERDICT line. (1) 'pnpm --filter @object-ui/plugin-charts type-check' + 'pnpm exec vitest run packages/plugin-charts/' gave: script name echoed (not a zero-match run), 'Test Files 32 passed (32)' / 'Tests 235 passed (235)', 'os-verify-lock: VERDICT command-exit 0'. (2) the five plugin-dashboard tests that import @object-ui/plugin-charts and render the real chart chain (DatasetWidget.chartConfig.dom, DatasetWidget.comboPresentation, DashboardGridLayout.sduiDatasetPath, DashboardGridLayout.datasetPath, StaticTableWidget) gave 'Test Files 5 passed (5)' / 'Tests 46 passed (46)', 'VERDICT command-exit 0'. (3) node scripts/check-changeset-presence.mjs: '2 source file(s) of 1 released package(s) changed, and this change declares 1 changeset(s)'. (4) node scripts/check-changeset-no-major.mjs: 'No changeset declares a major bump'. (5) npm run check:control-bytes: 'check-control-bytes: OK (scanned 4553 tracked text file(s); skipped 85 binary)'. (6) eslint over the whole affected package: 45 files counted from --format json, 0 errors; my two files 0 errors / 0 new warnings (the 6 on ChartContainerImpl.tsx are pre-existing no-explicit-any). Narrowing declared and evidenced: population read from eslint's own config resolution rather than a guess, count read from --format json, and no type-aware linting is configured (no projectService / parserOptions.project anywhere in the eslint configs), so this diff cannot move the verdict on any untouched file. REVERSE VERIFICATION, from a committed state: 'git checkout origin/main -- packages/plugin-charts/src/ChartContainerImpl.tsx', mutation confirmed on disk by grep counts (CHART_MIN_HEIGHT 0 occurrences, original 'minHeight: 280' text present) gave 'Tests 2 failed | 2 passed (4)', 'AssertionError: expected \"\" to be \"280px\"', 'VERDICT command-exit 1'. Predicted direction and observed direction agree: the 2 floor assertions go red while the 2 'author's explicit height wins' assertions stay green, which is required — the fix must not change that precedence. Restored with git checkout of the branch ref for that one path, restoration confirmed on disk (5 occurrences). No ablation of built dist was involved: the subject is source-level and the harness renders src directly. BROWSER before/after on the same rig, wrapper h-[350px] neutralised so the fallback is the only thing holding the box up — before: measured div 638x0, 0 surfaces, 0 marks; after: measured div 638x280, 2 surfaces, 3 sectors + 3 bars; untouched control unchanged at 638x350 with the same marks both times. DEVIATION (declared): a combined plugin-charts + plugin-dashboard vitest run acquired the lock after 8m48s and was killed by the container's 10-minute foreground cap, so I narrowed plugin-dashboard to the five files that actually import the chart package. The rest of plugin-dashboard is CI's run.",
  "open_questions": [
    {
      "question": "Is the maintainer's live-rig blank chart this permanent-zero-height mode, or a second, still-unfound cause? The measured symptom shape matches exactly (title renders, chart area blank, no message, pie and bar alike, no refusal), but the reported 'F5 paints it' asymmetry is not explained by it — a box that never changes is not fixed by a reload either.",
      "options": [
        "A: re-run the 2026-08-21 cloud-rig repro with PR #5503 applied. If the chart paints, this was the cause and the F5 note was a one-off observation; close #5451. Costs one rig run.",
        "B: keep #5451 open and dispatch a follow-up that captures the failing DOM from the real rig (the measured div's box and the computed height chain on the data-slot=chart element), which is the one datum no synthetic harness can supply. Costs a round but is decisive either way.",
        "C: treat the card as fixed on the strength of the symptom match alone. Cheapest, and wrong if a second cause exists — the symptom would come straight back."
      ],
      "recommendation": "A, then B only if it still reproduces. A is one run and answers the question directly; the fix is already independently justified by the measured before/after, so A risks nothing. C is not acceptable: I refuted the card's stated mechanism but did not reproduce its stated asymmetry, and closing on a symptom match would bury exactly that gap."
    }
  ],
  "out_of_scope_findings": [
    "filed as #5496 (sub-issue of #4695, label 'finding', unassigned, not queued): a chart whose declared series dataKey exists in NO row draws axes around zero marks and says nothing — the third silent-empty-frame arm, covered by neither #4683's 'series: []' guard nor #4695's 'series === undefined' arm. Measured, not inferred: rows carrying {status, count} against series [{dataKey:'id'}] produced 2 surfaces / 0 marks / 0 refusals; correcting the column made both charts paint."
  ]
}
```

---
_Generated by [Claude Code](https://claude.ai/code/session_014zHsbJoTkTZeJQ5DLbRXrE)_

---
_Generated by [Claude Code](https://claude.ai/code)_

--- a/.changeset/chart-height-chain-5451.md
+++ b/.changeset/chart-height-chain-5451.md
@@ -1,0 +1,5 @@
+---
+'@object-ui/plugin-charts': patch
+---
+
+`ObjectChart`'s wrapper div now carries `h-full`, keeping the height chain intact from a dashboard grid cell's declared height down to the element Recharts measures. Previously the chain died at the plain auto-height wrapper: `height: 100%` on the chart container computed to `auto`, Recharts measured a permanent zero, and only the `CHART_MIN_HEIGHT` floor (#5503) kept dashboard charts visible — at a fixed floor height instead of filling the cell (#5451). Under auto-height parents `h-full` resolves to `auto`, so non-dashboard hosts are unchanged.

--- a/packages/plugin-charts/src/ObjectChart.heightChain.test.tsx
+++ b/packages/plugin-charts/src/ObjectChart.heightChain.test.tsx
@@ -1,0 +1,57 @@
+/**
+ * ObjectUI
+ * Copyright (c) 2024-present ObjectStack Inc.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * objectui#5451 — the height chain through ObjectChart's wrapper.
+ *
+ * A dashboard grid cell declares a definite height and sends
+ * `className: "h-full"` down the widget schema; that class lands on
+ * ChartContainer, whose `height: 100%` resolves against ObjectChart's OWN
+ * wrapper div. When that wrapper is a plain auto-height block the chain dies
+ * there: the container computes to `auto`, recharts measures a permanent
+ * zero, and only the CHART_MIN_HEIGHT floor (#5503) keeps the chart from
+ * rendering blank — at a fixed floor height instead of filling the cell.
+ *
+ * Pinned here: the wrapper carries `h-full` so a parent-declared height
+ * actually reaches the measured element. (Under auto-height parents `h-full`
+ * resolves to `auto`, so non-dashboard hosts are unaffected — that is why
+ * this is safe unconditionally.)
+ */
+
+import React from 'react';
+import { describe, it, expect, vi, afterEach } from 'vitest';
+import { render, cleanup } from '@testing-library/react';
+
+vi.mock('./ChartRenderer', () => ({
+  ChartRenderer: () => null,
+}));
+
+import { ObjectChart } from './ObjectChart';
+
+afterEach(() => {
+  cleanup();
+});
+
+describe('ObjectChart wrapper height chain (objectui#5451)', () => {
+  it('the wrapper div propagates parent height (h-full), so a grid cell height reaches ChartContainer', () => {
+    const { container } = render(
+      <ObjectChart
+        schema={{
+          type: 'object-chart',
+          chartType: 'bar',
+          objectName: 'task',
+          xAxisKey: 'status',
+          className: 'h-full',
+          data: [{ status: 'open', count: 3 }],
+        }}
+        dataSource={{ find: async () => ({ data: [] }) }}
+      />,
+    );
+    const wrapper = container.firstElementChild as HTMLElement;
+    expect(wrapper).toBeTruthy();
+    expect(wrapper.className).toContain('h-full');
+  });
+});

--- a/packages/plugin-charts/src/ObjectChart.tsx
+++ b/packages/plugin-charts/src/ObjectChart.tsx
@@ -1000,7 +1000,14 @@ export const ObjectChart = (props: any) => {
   })() : null;
 
   return (
-    <div className="relative">
+    // `h-full` keeps the height chain intact: a dashboard grid cell declares a
+    // definite height and passes `className: "h-full"` down the schema to
+    // ChartContainer, whose `height: 100%` resolves against THIS div — a plain
+    // auto-height block here computes that to `auto`, recharts measures a
+    // permanent zero, and only the CHART_MIN_HEIGHT floor keeps the chart
+    // visible at all (#5451). Under auto-height parents `h-full` itself
+    // resolves to `auto`, so non-dashboard hosts are unchanged.
+    <div className="relative h-full">
       <RefreshIndicator active={loading && finalData.length > 0} />
       <ChartRenderer {...props} schema={finalSchemaWithColors} onChartClick={onChartClick} />
       {drillDrawer}

